### DISCUSSION TITLE
How does one go about populating an address book?

### DISCUSSION THREAD

**Discussion by windowsrefund** [the-djmaze/snappymail#812] [Q&A]
There doesn't seem to be any "Add to contacts" functionality anywhere? I do have the contacts database created and have verified connectivity. I also have the "Contacts from IMAP" extension (version 1.1) installed and working. With all that in place, I'd expect there to be some way to create a address book (or contact groups) based on results obtained from that plugin. Am I missing something obvious? Thanks in advance.

**Comment by the-djmaze** [comment]
RainLoop uses `IncFrec`
https://github.com/RainLoop/rainloop-webmail/blob/9bf8fa5b63d7e6871ee533126a7e1b1c85ee45b5/rainloop/v/0.0.0/app/libraries/RainLoop/Actions.php#L6447-L6469

This does two things when sending a message:
1. count how many times an email address is send to
2. if 0 add it to Contacts when setting ContactsAutosave (/#/settings/contacts or application.ini contacts_autosave) is true

SnappyMail only does step 2 at
https://github.com/the-djmaze/snappymail/blob/56c681bbd54294717c00cf7f75990347fa85f377/snappymail/v/0.0.0/app/libraries/RainLoop/Actions/Messages.php#L271-L286

BUT... The support of IncFrec() depends on the AddressBookInterface implementation, like at:
https://github.com/the-djmaze/snappymail/blob/56c681bbd54294717c00cf7f75990347fa85f377/snappymail/v/0.0.0/app/libraries/RainLoop/Providers/AddressBook/PdoAddressBook.php#L939-L943
and
https://github.com/the-djmaze/snappymail/blob/56c681bbd54294717c00cf7f75990347fa85f377/plugins/kolab/KolabAddressBook.php#L354-L357

So best is to check your setup if it does add the contact to AddressBook when you send a message.

Additionally it would indeed be nice to add contacts from message menu.

**Comment by windowsrefund** [comment]
Right, it does add on send but that's still essentially having to (manually) recreate a database; 1 record at a time. We're overlooking the opportunity to leverage the inbox as a source of truth.
